client/state/happychat/chat/reducer.js

```diff
@@ -0,0 +1,181 @@
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ *
+ */
+export const message = ( state = '', action ) => {
+	switch ( action.type ) {
+		case HAPPYCHAT_SEND_MESSAGE:
+			return '';
+		case HAPPYCHAT_SET_MESSAGE:
+			return action.message;
+	}
+	return state;
+};
+
+/**
+ * Returns a timeline event from the redux action
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ *
+ */
+const timelineEvent = ( state = {}, action ) => {
+	switch ( action.type ) {
+		case HAPPYCHAT_RECEIVE_EVENT:
+			const event = action.event;
+			return Object.assign(
+				{},
+				{
+					id: event.id,
+					source: event.source,
+					message: event.text,
+					name: event.user.name,
+					image: event.user.avatarURL,
+					timestamp: event.timestamp,
+					user_id: event.user.id,
+					type: get( event, 'type', 'message' ),
+					links: get( event, 'meta.links' ),
+				}
+			);
+	}
+	return state;
+};
+
+const validateTimeline = validator( timelineSchema );
+const sortTimeline = timeline => sortBy( timeline, event => parseInt( event.timestamp, 10 ) );
+
+/**
+ * Adds timeline events for happychat
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ *
+ */
+export const timeline = ( state = [], action ) => {
+	switch ( action.type ) {
+		case SERIALIZE:
+			return takeRight( state, HAPPYCHAT_MAX_STORED_MESSAGES );
+		case DESERIALIZE:
+			const valid = validateTimeline( state );
+			if ( valid ) {
+				return state;
+			}
+			return [];
+		case HAPPYCHAT_RECEIVE_EVENT:
+			// if meta.forOperator is set, skip so won't show to user
+			if ( get( action, 'event.meta.forOperator', false ) ) {
+				return state;
+			}
+			const event = timelineEvent( {}, action );
+			const existing = find( state, ( { id } ) => event.id === id );
+			return existing ? state : concat( state, [ event ] );
+		case HAPPYCHAT_TRANSCRIPT_RECEIVE:
+			const messages = filter( action.messages, msg => {
+				if ( ! msg.id ) {
+					return false;
+				}
+
+				// if meta.forOperator is set, skip so won't show to user
+				if ( get( message, 'meta.forOperator', false ) ) {
+					return false;
+				}
+
+				return ! find( state, { id: message.id } );
+			} );
+			return sortTimeline(
+				state.concat(
+					map( messages, msg => {
+						return Object.assign( {
+							id: msg.id,
+							source: msg.source,
+							message: msg.text,
+							name: msg.user.name,
+							image: msg.user.picture,
+							timestamp: msg.timestamp,
+							user_id: msg.user.id,
+							type: get( msg, 'type', 'message' ),
+							links: get( msg, 'meta.links' ),
+						} );
+					} )
+				)
+			);
+	}
+	return state;
+};
+timeline.hasCustomPersistence = true;
+
+export default combineReducers( {
+	status,
+	timeline,
+	message,
```

Comment:
I think there's an argument to be made that `message` should actually be under `happychat.ui` instead of `happychat.chat`. This field is only used within the Composer to track the message currently being typed by the user and isn't intrinsically tied to the `chat` itself in any way.

Reply:
I like this suggestion: 68458c6c8c35b45aaffabc9271d5f2aa2420266a